File: js-port/funcscript-js/src/block/language-binding-block.js
const { ExpressionBlock } = require('./expression-block');
const { assertTyped, normalize } = require('../core/value');
const { FsError } = require('../model/fs-error');

class LanguageBindingBlock extends ExpressionBlock {
  constructor(languageIdentifier, code, binding, position = 0, length = 0) {
    super(position, length);
    if (!binding) {
      throw new Error('Language binding is required.');
    }

    this._languageIdentifier = languageIdentifier || '';
    this._code = code || '';
    this._binding = binding;

    let compiled = null;
    let compileError = null;
    try {
      const result = binding.compile(this._code);
      if (result && typeof result === 'object') {
        compiled = result.compiled ?? result.Compiled ?? null;
        compileError =
          typeof result.error === 'string'
            ? result.error
            : typeof result.Error === 'string'
            ? result.Error
            : null;
      } else {
        compiled = result;
      }
    } catch (error) {
      compileError = error && error.message ? error.message : `Failed to compile '${this._languageIdentifier}'.`;
    }

    this._compiledCode = compiled;
    this._compileError = compileError;
  }

  evaluateInternal(provider) {
    if (this._compileError) {
      const message = this._languageIdentifier
        ? `[${this._languageIdentifier}] ${this._compileError}`
        : this._compileError;
      return normalize(new FsError(FsError.ERROR_DEFAULT, message));
    }

    try {
      const result = this._binding.evaluate(this._compiledCode, provider);
      return assertTyped(result);
    } catch (error) {
      const message =
        (error && error.message) || `Language binding '${this._languageIdentifier || 'unknown'}' evaluation failed.`;
      return normalize(new FsError(FsError.ERROR_DEFAULT, message));
    }
  }

  asExpressionString() {
    const escapedCode = (this._code || '').replace(/```/g, '\\```');
    const lang = this._languageIdentifier || '';
    return `\`\`\`${lang}\n${escapedCode}\n\`\`\``;
  }
}

module.exports = {
  LanguageBindingBlock
};

File: FuncScript/Block/LanguageBindingBlock.cs
using System;
using FuncScript.Core;
using FuncScript.Error;
using FuncScript.Model;

namespace FuncScript.Block
{
    public class LanguageBindingBlock : ExpressionBlock
    {
        private readonly string _languageIdentifier;
        private readonly string _code;
        private readonly ILanguageBinding _binding;
        private readonly object _compiledCode;
        private readonly string _compileError;

        public LanguageBindingBlock(string languageIdentifier, string code, ILanguageBinding binding)
        {
            _languageIdentifier = languageIdentifier ?? throw new ArgumentNullException(nameof(languageIdentifier));
            _code = code ?? string.Empty;
            _binding = binding ?? throw new ArgumentNullException(nameof(binding));

            CompilationResult compilation;
            try
            {
                compilation = _binding.Compile(_code);
            }
            catch (Exception ex)
            {
                var message = string.IsNullOrWhiteSpace(ex.Message)
                    ? $"Failed to compile language binding '{_languageIdentifier}'."
                    : ex.Message;
                compilation = new CompilationResult(null, message);
            }

            _compiledCode = compilation.Compiled;
            _compileError = compilation.Error;
        }

        public override object Evaluate(KeyValueCollection provider,DepthCounter depth)
        {
            var entryState = depth.Enter(this);
            object result = null;
            try
            {

                if (!string.IsNullOrEmpty(_compileError))
                {
                    var message = string.IsNullOrWhiteSpace(_languageIdentifier)
                        ? _compileError
                        : $"[{_languageIdentifier}] {_compileError}";
                    result = AttachCodeLocation(this, new FsError(FsError.ERROR_DEFAULT, message));
                    return result;
                }

                result = _binding.Evaluate(_compiledCode, provider);
                if (result is FsError bindingError)
                    result = AttachCodeLocation(this, bindingError);
                return result;
            }
            catch (EvaluationException)
            {
                throw;
            }
            catch (Exception ex)
            {
                var message = string.IsNullOrWhiteSpace(ex.Message)
                    ? $"Language binding '{_languageIdentifier}' evaluation failed."
                    : ex.Message;
                result = AttachCodeLocation(this, new FsError(FsError.ERROR_DEFAULT, message));
                return result;
            }
            finally
            {
                depth.Exit(entryState, result, this);
            }
        }

        public override string AsExpString()
        {
            var escapedCode = (_code ?? string.Empty).Replace("```", "\\```");
            return $"```{_languageIdentifier}\n{escapedCode}\n```";
        }
    }
}
